### PR TITLE
Including jvp for radius update schemes

### PR DESCRIPTION
**Pull request by yash2798**

### CODE REVIEW

File: src/jacobian.jl [hunk 1]
```diff
@@ -144,3 +144,14 @@ function jacobian_autodiff(f, x::AbstractArray, nonlinfun, alg)
                                 jac_prototype = jac_prototype, chunksize = chunk_size),
      num_of_chunks)
 end
+
+function jvp(cache::TrustRegionCache{false})
```
Comment (ChrisRackauckas) [inline src/jacobian.jl:148]: Put a `!` on there because it's mutating.

File: src/trustRegion.jl [hunk 2]
```diff
@@ -417,17 +417,18 @@ function trust_region_step!(cache::TrustRegionCache)
...
         take_step!(cache)
         cache.loss = cache.loss_new
         cache.make_new_J = true
       else
         cache.make_new_J = false
       end
-      
+
+      @unpack p1= cache
+      cache.trust_r = p1 * cache.internalnorm(jvp(cache)) # we need the gradient at the new (k+1)th point  WILL THIS BECOME ALLOCATING?
```
Comment (ChrisRackauckas) [inline src/trustRegion.jl:430]: The comment can be removed.

**Comment by ChrisRackauckas** [comment]
Needs to add the dependency.

File: src/jacobian.jl (hunk 1 above)
Comment (yash2798) [reply src/jacobian.jl:148]: we are actually not modifying the cache, we are just returning the jvp. Should that still be done?

File: src/NonlinearSolve.jl [hunk 3]
```diff
@@ -61,6 +61,7 @@ SnoopPrecompile.@precompile_all_calls begin for T in (Float32, Float64)
 end end
 
 export RadiusUpdateSchemes
+export auto_jacvec, auto_jacvec!
```
Comment (ChrisRackauckas) [inline src/NonlinearSolve.jl:64]:
```suggestion
```

File: src/jacobian.jl (hunk 1 above)
Comment (ChrisRackauckas) [reply src/jacobian.jl]: You are modifying the cache in one of them

File: src/trustRegion.jl [hunk 4]
```diff
@@ -441,7 +445,7 @@ function dogleg!(cache::TrustRegionCache)
 
     # Test if the full step is within the trust region.
     if norm(u_tmp) ≤ trust_r
-        cache.step_size = u_tmp
+        cache.step_size = deepcopy(u_tmp)
```
Comment (ChrisRackauckas) [inline src/trustRegion.jl:448]: Is this supposed to be a `.=`?

**Comment by yash2798** [comment]
The issue is resolved, if there's nothing else, i think we can merge this pr.

**Comment by ChrisRackauckas** [comment]
Test failed, looks like a tolerance issue.

**Comment by yash2798** [comment]
works now

**Review by ChrisRackauckas — APPROVED** [review]

**Pull request merged**